File: gridgo-server/app/schemas/room.py
"""
房间相关的 Pydantic 模型
"""

from datetime import datetime

from pydantic import BaseModel, Field


# ─── 请求模型 ───


class CreateRoomRequest(BaseModel):
    """创建房间请求"""

    name: str = Field(default="GridGo 房间", max_length=50, description="房间名称")
    max_players: int = Field(default=8, ge=2, le=8, description="最大玩家数")
    map_id: str = Field(default="classic", description="地图 ID")
    ai_count: int = Field(default=0, ge=0, le=8, description="AI 数量")
    ai_difficulty: str = Field(default="easy", description="AI 难度: easy/medium/hard")

    def model_post_init(self, __context) -> None:
        """校验 ai_count 不超过 max_players"""
        if self.ai_count > self.max_players:
            raise ValueError(f"AI 数量({self.ai_count})不能超过最大玩家数({self.max_players})")


class UpdateRoomRequest(BaseModel):
    """更新房间设置请求（仅房主）"""

    name: str | None = Field(default=None, max_length=50, description="房间名称")
    max_players: int | None = Field(default=None, ge=2, le=8, description="最大玩家数")
    map_id: str | None = Field(default=None, description="地图 ID")
    ai_count: int | None = Field(default=None, ge=0, le=8, description="AI 数量")
    ai_difficulty: str | None = Field(default=None, description="AI 难度: easy/medium/hard")


class JoinRoomRequest(BaseModel):
    """加入房间请求"""

    code: str = Field(..., min_length=6, max_length=6, description="6位房间代码")
    as_spectator: bool = Field(default=False, description="以观战者身份加入")


class SwitchRoleRequest(BaseModel):
    """切换角色请求（玩家 ↔ 观战者）"""

    to_spectator: bool = Field(..., description="True=切换为观战者，False=切换为玩家")


class SpectatorInfo(BaseModel):
    """观战者信息"""

    user_id: int
    username: str
    nickname: str
    avatar: str | None = None
    is_host: bool = False  # 房主也可以是观战者


# ─── 响应模型 ───


class RoomPlayer(BaseModel):
    """房间中的玩家"""

    user_id: int
    username: str
    nickname: str
    avatar: str | None = None
    is_host: bool = False
    is_ready: bool = False
    is_ai: bool = False
    ai_difficulty: str | None = None


class RoomInfo(BaseModel):
    """房间信息"""

    id: str
    code: str
    name: str
    host_id: int
    max_players: int
    map_id: str = "classic"
    ai_count: int = 0
    ai_difficulty: str = "easy"
    status: str  # waiting / playing / finished
    players: list[RoomPlayer] = []
    spectators: list[SpectatorInfo] = []  # 观战者列表
    max_spectators: int = 10  # 最大观战人数
    created_at: datetime | None = None


class RoomListItem(BaseModel):
    """房间列表项（简要信息）"""

    id: str
    code: str
    name: str
    host_nickname: str
    player_count: int
    max_players: int
    map_id: str = "classic"
    ai_count: int = 0
    spectator_count: int = 0  # 观战人数
    status: str

File: gridgo-web/src/api/room.ts
/**
 * 房间相关 API
 */

import request from './rest'

// ─── 类型定义 ───

export interface RoomPlayer {
  user_id: number
  username: string
  nickname: string
  avatar: string | null
  is_host: boolean
  is_ready: boolean
  is_ai: boolean
  ai_difficulty: string | null
}

export interface SpectatorInfo {
  user_id: number
  username: string
  nickname: string
  avatar: string | null
  is_host: boolean
}

export interface RoomInfo {
  id: string
  code: string
  name: string
  host_id: number
  max_players: number
  map_id: string
  ai_count: number
  ai_difficulty: string
  status: 'waiting' | 'playing' | 'finished'
  players: RoomPlayer[]
  spectators: SpectatorInfo[]
  max_spectators: number
  /** 是否设置了房间密码（后端 RoomInfo.has_password） */
  has_password?: boolean
  created_at: string | null
}

export interface RoomListItem {
  id: string
  code: string
  name: string
  host_nickname: string
  player_count: number
  max_players: number
  map_id: string
  ai_count: number
  spectator_count: number
  status: string
}

export interface CreateRoomData {
  name?: string
  max_players?: number
  map_id?: string
  ai_count?: number
  ai_difficulty?: string
  /** 房间密码（可选，设置后加入房间需提供） */
  password?: string | null
}

export interface UpdateRoomData {
  name?: string
  max_players?: number
  map_id?: string
  ai_count?: number
  ai_difficulty?: string
  /** 房间密码：传值表示设置/修改；配合 clear_password 可清除 */
  password?: string | null
  /** 清除房间密码 */
  clear_password?: boolean
}

// ─── 地图选项 ───

export const MAP_OPTIONS = [
  { value: 'classic', label: '经典地图' },
  { value: 'city', label: '城市风云' },
  { value: 'island', label: '海岛探险' },
] as const

export const AI_DIFFICULTY_OPTIONS = [
  { value: 'easy', label: '简单' },
  { value: 'medium', label: '中等' },
  { value: 'hard', label: '困难' },
] as const

// ─── API 方法 ───

/** 获取房间列表 */
export function getRooms() {
  return request.get<any, RoomListItem[]>('/rooms')
}

/** 获取当前用户所在（活跃）房间；未加入任何房间时返回 null */
export function getMyRoom() {
  return request.get<any, RoomInfo | null>('/rooms/active')
}

/** 创建房间 */
export function createRoom(data?: CreateRoomData) {
  return request.post<any, RoomInfo>('/rooms', data || {})
}

/** 通过房间代码获取房间 */
export function getRoomByCode(code: string) {
  return request.get<any, RoomInfo>(`/rooms/by-code/${code}`)
}

/** 获取房间详情 */
export function getRoom(roomId: string) {
  return request.get<any, RoomInfo>(`/rooms/${roomId}`)
}

/** 加入房间（密码房需传 password） */
export function joinRoom(code: string, asSpectator: boolean = false, password?: string) {
  const body: { code: string; as_spectator: boolean; password?: string } = {
    code,
    as_spectator: asSpectator,
  }
  if (password) body.password = password
  return request.post<any, RoomInfo>('/rooms/join', body)
}

/** 离开房间 */
export function leaveRoom(roomId: string) {
  return request.post<any, { code: number; destroyed: boolean; message: string }>(`/rooms/${roomId}/leave`)
}

/** 准备/取消准备 */
export function toggleReady(roomId: string) {
  return request.post<any, RoomInfo>(`/rooms/${roomId}/ready`)
}

/** 更新房间设置 */
export function updateRoomSettings(roomId: string, data: UpdateRoomData) {
  return request.patch<any, RoomInfo>(`/rooms/${roomId}/settings`, data)
}

/** 开始游戏 */
export function startGame(roomId: string) {
  return request.post<any, RoomInfo>(`/rooms/${roomId}/start`)
}

/** 切换角色（玩家 ↔ 观战者） */
export function switchRole(roomId: string, toSpectator: boolean) {
  return request.post<any, RoomInfo>(`/rooms/${roomId}/switch-role`, { to_spectator: toSpectator })
}

/** 房主重置房间（游戏结束后返回准备阶段） */
export function resetRoom(roomId: string) {
  return request.post<any, RoomInfo>(`/rooms/${roomId}/reset`)
}
